File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters/Session/ChangeSet.cs
// <copyright file="ChangeSet.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
// <summary>
//   Defines the AllorsChangeSetMemory type.
// </summary>

namespace Allors.Workspace.Adapters
{
    using System.Collections.Generic;
    using System.Linq;
    using Meta;

    public sealed class ChangeSet : IChangeSet
    {
        public ChangeSet(Session session, ISet<IStrategy> created, ISet<IStrategy> instantiated)
        {
            this.Session = session;
            this.Created = created;
            this.Instantiated = instantiated;
            this.AssociationsByRoleType = new Dictionary<IRoleType, ISet<IStrategy>>();
            this.RolesByAssociationType = new Dictionary<IAssociationType, ISet<IStrategy>>();
        }

        private Session Session { get; }

        ISession IChangeSet.Session => this.Session;

        public ISet<IStrategy> Created { get; }

        public ISet<IStrategy> Instantiated { get; }

        public IDictionary<IRoleType, ISet<IStrategy>> AssociationsByRoleType { get; }

        public IDictionary<IAssociationType, ISet<IStrategy>> RolesByAssociationType { get; }

        public void AddSessionStateChanges(
            IDictionary<IPropertyType, IDictionary<long, object>> sessionStateChangeSet)
        {
            foreach (var kvp in sessionStateChangeSet)
            {
                var ids = kvp.Value.Keys;
                var strategies = new HashSet<IStrategy>(ids.Select(v => this.Session.GetStrategy(v)));

                switch (kvp.Key)
                {
                    case IAssociationType associationType:
                        this.RolesByAssociationType.Add(associationType, strategies);
                        break;
                    case IRoleType roleType:
                        this.AssociationsByRoleType.Add(roleType, strategies);
                        break;
                }
            }
        }

        public void Diff(Strategy association, IRelationType relationType, object current, object previous)
        {
            var roleType = relationType.RoleType;

            if (roleType.ObjectType.IsUnit)
            {
                if (!Equals(current, previous))
                {
                    this.AddAssociation(relationType, association);
                }
            }
            else
            {
                if (roleType.IsOne)
                {
                    if (Equals(current, previous))
                    {
                        return;
                    }

                    if (previous != null)
                    {
                        this.AddRole(relationType, this.Session.GetStrategy((long)previous));
                    }

                    if (current != null)
                    {
                        this.AddRole(relationType, this.Session.GetStrategy((long)current));
                    }

                    this.AddAssociation(relationType, association);
                }
                else
                {
                    var numbers = this.Session.Workspace.Database.Configuration.Numbers;
                    var hasChange = false;

                    var addedRoles = numbers.Except(current, previous);
                    foreach (var v in numbers.Enumerate(addedRoles))
                    {
                        this.AddRole(relationType, this.Session.GetStrategy(v));
                        hasChange = true;
                    }

                    var removedRoles = numbers.Except(previous, current);
                    foreach (var v in numbers.Enumerate(removedRoles))
                    {
                        this.AddRole(relationType, this.Session.GetStrategy(v));
                        hasChange = true;
                    }

                    if (hasChange)
                    {
                        this.AddAssociation(relationType, association);
                    }
                }
            }
        }

        private void AddAssociation(IRelationType relationType, Strategy association)
        {
            var roleType = relationType.RoleType;

            if (!this.AssociationsByRoleType.TryGetValue(roleType, out var associations))
            {
                associations = new HashSet<IStrategy>();
                this.AssociationsByRoleType.Add(roleType, associations);
            }

            _ = associations.Add(association);
        }

        private void AddRole(IRelationType relationType, Strategy role)
        {
            var associationType = relationType.AssociationType;

            if (!this.RolesByAssociationType.TryGetValue(associationType, out var roles))
            {
                roles = new HashSet<IStrategy>();
                this.RolesByAssociationType.Add(associationType, roles);
            }

            _ = roles.Add(role);
        }
    }
}

File: typescript/libs/workspace/domain/json/rxjs/system/src/lib/session/ChangeSet.ts
import { IChangeSet, IStrategy } from '@allors/workspace/system';
import { AssociationType, PropertyType, RelationType, RoleType } from '@allors/workspace/system';
import { Session } from './Session';
import { Strategy } from './Strategy';

export class ChangeSet implements IChangeSet {
  associationsByRoleType: Map<RoleType, Set<IStrategy>>;

  rolesByAssociationType: Map<AssociationType, Set<IStrategy>>;

  public constructor(public Session: Session, public Created: Set<IStrategy>, public Instantiated: Set<IStrategy>) {
    this.associationsByRoleType = new Map();
    this.rolesByAssociationType = new Map();
  }

  public AddSessionStateChanges(sessionStateChangeSet: Map<PropertyType, Map<number, any>>) {
    sessionStateChangeSet.forEach((map, propertyType) => {
      const ids = map.keys;

      const strategies = new Set<IStrategy>(this.Session.GetStrategies(ids));

      if (propertyType.isAssociationType) {
        this.rolesByAssociationType.set(propertyType as AssociationType, strategies);
      } else if (propertyType.isRoleType) {
        this.associationsByRoleType.set(propertyType as RoleType, strategies);
      }
    });
  }

  public Diff(association: Strategy, relationType: RelationType, current: any, previous: any) {
    const roleType = relationType.roleType;
    if (roleType.objectType.isUnit) {
      if (current !== previous) {
        this.AddAssociation(relationType, association);
      }
    } else if (roleType.isOne) {
      if (current === previous) {
        return;
      }

      if (previous != null) {
        this.AddRole(relationType, this.Session.getStrategy(<number>previous));
      }

      if (current != null) {
        this.AddRole(relationType, this.Session.getStrategy(<number>current));
      }

      this.AddAssociation(relationType, association);
    } else {
      const numbers = this.Session.workspace.Numbers;
      let hasChange = false;
      const addedRoles = numbers.Except(current, previous);
      for (const v in numbers.Enumerate(addedRoles)) {
        this.AddRole(relationType, this.Session.getStrategy(v));
        hasChange = true;
      }

      const removedRoles = numbers.Except(previous, current);
      for (const v in numbers.Enumerate(removedRoles)) {
        this.AddRole(relationType, this.Session.getStrategy(v));
        hasChange = true;
      }

      if (hasChange) {
        this.AddAssociation(relationType, association);
      }
    }
  }

  private AddAssociation(relationType: RelationType, association: Strategy) {
    const roleType = relationType.roleType;

    let associations: Set<IStrategy>;
    if (!this.associationsByRoleType.has(roleType)) {
      associations = new Set();
      this.associationsByRoleType.set(roleType, associations);
    }

    associations.add(association);
  }

  private AddRole(relationType: RelationType, role: Strategy) {
    const associationType = relationType.AssociationType;

    let roles: Set<IStrategy>;
    if (!this.rolesByAssociationType.has(associationType)) {
      roles = new Set();
      this.rolesByAssociationType.set(associationType, roles);
    }

    roles.add(role);
  }
}
